File: src/p2xxx/p2563.py
from leetcode_prelude import *


# Problem 2563
class Solution:
    def countFairPairs(self, nums: List[int], lower: int, upper: int) -> int:
        """
        1. 0 <= i < j < n imply that a pair is formed in combination sense
        2. we may sort the array in this sense
        3. two pointer approach to count sums less than or equal to a value
        4. the difference between two counts is the answer
        """
        if len(nums) < 2:
            return 0

        nums.sort()

        def count_le(value: int) -> int:
            count = 0
            left = 0
            right = len(nums) - 1

            while left < right:
                while left < right and nums[left] + nums[right] > value:
                    right -= 1

                count += right - left
                left += 1

            return count

        return count_le(upper) - count_le(lower - 1)

File: src/p2xxx/p2563.rs
// Problem 2563
impl Solution {
    pub fn count_fair_pairs(mut nums: Vec<i32>, lower: i32, upper: i32) -> i64 {
        nums.sort_unstable();

        fn count_le(nums: &[i32], bound: i32) -> i64 {
            let mut count = 0;
            let mut left = 0;
            let mut right = nums.len() - 1;

            while left < right {
                while left < right && nums[left] + nums[right] > bound {
                    right -= 1;
                }

                count += right - left;
                left += 1;
            }

            count as i64
        }

        count_le(&nums, upper) - count_le(&nums, lower - 1)
    }
}
